### ISSUE TITLE
How to handle “post-game” judge re-assessments

### ISSUE THREAD

**Issue by CloudEllie** [code-423n4/org#12]
## Problem

For contests that were judged prior to our instatement of a 48-hour sponsor review period (during which sponsors run a QA check on the judge’s assessments), we’ve had several cases where the sponsor and judge revisit issues **after** awards have been calculated, announced, and added to the leaderboard, and either downgrade or upgrade severity.

While we have instituted some process changes that should reduce the frequency with which this happens, there's always some risk of "blown calls," where new context comes to light after the judging period has ended, and in the spirit of fairness, there is a desire to re-assess decisions (missed duplicates, flubbed severity ratings, etc.). 

## How we’ve handled it so far

In an attempt to a) accommodate sponsors & judges’ desires, and b) ensure our reports are as accurate as possible, I (Ellie) have taken the following approach to date:

- Update findings sheet;
- Re-run award calculations;
- Determine which wardens were owed *more* than they were originally paid out;
- Send additional payouts to those wardens;
- Update the leaderboard with the corrected awardcalc results.

## Upsides of past approach

- Leaderboard more accurately reflects wardens’ numbers of High & Medium issues found;
- Any wardens who were owed more money hopefully gain a sense that C4 rewards people fairly.

## Downsides of past approach

This is a painstaking, laborious process and has some negative effects:

- Costs C4 money in additional warden payouts;
- Increases the discrepancies between actual payouts vs what’s displayed on the leaderboard;
- Makes some wardens unhappy, because their leaderboard numbers can go down (e.g. downgraded issues, and/or the awardcalc re-run results in lower awards shown on leaderboard).

## Questions

- What might a better solution to the above-stated problem look like?
- What should our policy be, in cases where a judge and sponsor agree that an issue ought to be re-assessed “after the fact”?

## Proposal

I propose that we take the following approach (and establish a policy) for contests where this arises, going forward:

1. Once awards have been announced, we will make no further adjustments to warden payouts or leaderboard scoring.
2. We *may* adjust the risk levels of issues included in the report, but will include a note to indicate that the issue(s) in question were re-assessed by the judge and/or sponsor after the judge’s decisions were announced. 
3. In extraordinary cases, where a warden’s submission was unfairly excluded from duly-owed awards:
    1. Judges can propose a modest ARENA token grant (say, 100% of the award difference, up to a max of $5k) to make up for blown calls (measured in USD value, not ARENA count), and we batch these together periodically when we do airdrops.
    2. However, C4 will NOT update the leaderboard scoring.

**Comment by IllIllI000** [comment]
The above proposal solves the cost to C4, but it doesn't solve discrepancies if they're still compensated with ARENA, and I don't think many wardens would be all that unhappy about any finding being marked how it should have been in the first place, especially if nobody gets less actual money than before. The painstaking and laborious process still would have to be done for those cases too, so there's not less work to be done in those cases either.

I think wardens care much more about actual rewards received than of potential changes to the leaderboard, so I would hope that the solution chosen favors that aspect wherever possible. Also, if there's no penalty to the judge who called things incorrectly, there's no incentive to improve. How about:

Whenever there's a sufficient number of complaints about a call (not sure what the lower bound is, but obvious cases will be easy to see in discord comments), have a vote by all of the other judges on what the correct call should have been. If 2/3 of the responding judges agree that it should have been called differently, the judge that got the call wrong has to do the following in a pull request:
1) Update the findings sheet
2) Run the rewards calculations
3) Determine who is owed more than they were originally paid, and outline that in the request comments
4) Update the leaderboard
C4 would immediately reimburse the wardens who were underpaid and then create a debt for the wrong judge. Until the debt is repaid, any contests that the judge judges or participates in as a warden (or has not yet been paid out), has max($X000,Y%) of the winnings/fees go towards repaying the debt.

C4 only loses money if the judge quits, which would incentivize C4 to choose good judges who also seem likely to stick around; the judge is forced to do the extra work and is incentivized to improve; the leaderboard is corrected; and wardens are happy about getting paid. There may be some up-front cost in setting up the voting, pull request, and debt systems, but once done they'd just work. I think the real extra work would be getting the judges to vote, but perhaps that could be incentivized with quarterly debts to judges who don't respond to all vote requests during the quarter, or something.

**Comment by TomAFrench** [comment]
>Judges can propose a modest ARENA token grant (say, 100% of the award difference, up to a max of $5k) to make up for blown calls (measured in USD value, not ARENA count), and we batch these together periodically when we do airdrops.

I think it would be good to clarify whether this grant is capped on a "per blown call" or "per contest" basis. Also is there a set max period for these airdrops? afaik there hasn't been one since the DAO formation and I'd hate to be left waiting for months in order to receive these funds.

**Comment by TomAFrench** [comment]
Other than this I think this proposal is very reasonable.

Is it that burdensome to update the leaderboard with the extra payout to the warden? I imagine the issue currently is from having to reduce the payouts for other wardens. Personally I wouldn't really care about others in that contest having an "incorrectly" high leaderboard position as long as I had the extra payout included in my total. Would it be that difficult to add these extra payouts to the leaderboard?

I strongly oppose @IllIllI000's debt based proposal as I think it'll have the counterproductive effect of incentivising judges not adequately rejecting findings with inflated severity ratings in order to avoid the potential for losing several thousand dollars.

That said I'd be interested in seeing if the certified warden program renders this proposal obsolete. Wardens often want to look over the report to find any issues they have missed and so would be well placed to review potentially invalid findings although this has potential for conflicts of interest.

**Comment by IllIllI000** [comment]
The debt in that sort of case would still apply since the people who got less because of an inflated severity finding would complain and would be paid

**Comment by TomAFrench** [comment]
I don't see the difference from the current situation then except shifting the shortfall from c4 onto judges who are arguably less able to bear this cost. afaik judges are part of the current bottleneck and so discouraging them from taking part wouldn't be helpful.

**Comment by IllIllI000** [comment]
It's a solution to the three stated downsides of the current approach while keeping the two stated upsides

**Comment by CloudEllie** [comment]
I just want to note here -- since this proposal has functionally become C4 policy, at least at a working level -- that now that we have Certified+ wardens reviewing the judges' work prior to calculating awards, we will not be accepting further "post-game" appeals from wardens. 

Going forward, we expect any QA concerns to be raised by Certified+ wardens during the post-judging QA period, so that the judge can consider them prior to finalizing their decisions. 

In other words: if a contest has gone through a post-judging QA period with Certified+ wardens, we will consider all judges' decisions final.

**Comment by kartoonjoy** [comment]
Per the Autumn 2023 C4 Supreme Court verdicts, the Supreme Court's verdict on this issue is: 

> No ruling changes are desirable.
> 
> `Recommendations:` We recommend improving the robustness of the PJQA stage through:
> An automatic changelog for judge verdicts
> A structured process to validate a request for review has been processed
> Exclusively for issues that were raised during the original PJQA timeline, a change or request for comments from the Judge allows an additional 24 hours of time for Wardens to raise additional concerns.
> 
> 
> Link to verdict:  https://docs.google.com/document/d/1Y2wJVt0d2URv8Pptmo7JqNd0DuPk_qF9EPJAj3iSQiE/edit#heading=h.kwdb1q7esdp1
